rfcs/0017-state-snapshotting.md

```diff
@@ -0,0 +1,244 @@
+person3:     C--------
+person4: D---
+person5: E--
+person6:              F--
+```
+
+### Full-offline Snapshotting
+
+This is an extension of partial snapshotting. In addition to all that is
+required above:
+
+The following is what is believed to be an exhaustive list of the pieces of
+state needed for full-offline snapshotting:
+
+1. Transition frontier + extensions
+
+For marshalling and unmarshalling of the data:
+
+We can rely on https://github.com/CodaProtocol/coda/pull/1779 (which should be
+worked on in parallel of partial snapshotting above).
+
+To load the snapshot record (in addition to the above):
+
+1. We start transition router with this particular transition frontier +
+extensions and disable all the pipes that trigger changes to the transition
+frontier system.
+
+2. We ensure that transaction snark work and transactions can be generated and
+sent to this node (despite all network activity being disable).
+
+For implementation we can do the following tasks:
+
+A. All of partial snapshotting
+
+B. Transition frontier persistance
+
+C. Modifying transition router and other parts of the code to properly load
+from snapshot
+
+D. Test to ensure multiple transitions are proposed successfully with new txns
+and snark works/proofs
+
+Parallelism of tasks with rough approximations of length:
+
+```
+group1:  A----------
+group2:  B----------
+person1:            C-----
+person2:                  D--
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+* These tasks would take quite a bit of effort to complete
+* The snapshotting system itself will be complex enough to have it's own bugs
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Alternatively, we could do something more quickly, but in a way that is much
+less flexible by doing something like taking a heap snapshot at a certain time
+and resuming from that point. We could even do this by snapshotting virtual
+machine memory. Such an approach, is not too useful; however, as we can't easily
+iterate on versions of the coda daemon while reusing the same snapshot, nor is
+it easy to resume such a snapshot that is taken on a larger testnet.
+
+The impact of not implementing this, is twofold (1) it takes us longer to
+reproduce errors we run into while testing and (2) longer to iterate on those
+error cases when we're actually fishing for a bug.
+
+## Prior art
+[prior-art]: #prior-art
+
+I could not find any prior art in the Bitcoin space. It seems like pyethereum
+and parity both support some form of state snapshotting to JSON. It's not clear
+to me that this state captures enough of the information necessary to resume a
+network from that point (and we can't dig through the source due to licensing
+issues), but if it does, it's nice that they managed to use JSON for their
+output format as it's cross-client compatible and easily consumable by other
+tooling.
+
+Some of the data we need to serialize would be a bit hard and slow if we put it
```

Comment:
I've been putting some of our data structures into a json format for our visualizations :D.